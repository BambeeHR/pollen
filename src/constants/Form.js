const Form = Object.freeze({
  Variants: {
    STANDARD: 'standard',
    RAISED: 'raised',
    // GHOST: 'ghost',
    // GHOST_INVERTED: 'ghost-inverted',
  },
  Sizes: {
    DENSE: 'dense',
    NORMAL: 'normal',
    LARGE: 'large',
  },
  LabelPositions: {
    LEFT: 'left',
    RIGHT: 'right',
  },
});

export default Form;
