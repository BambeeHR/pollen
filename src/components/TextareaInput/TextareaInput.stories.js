import capitalize from 'lodash/capitalize';
import { action } from '@storybook/addon-actions';
import { boolean, select, text } from '@storybook/addon-knobs';
import TextareaInput from './TextareaInput.vue';
import Form from '../../constants/Form';
import TypeOverline from '../TypeOverline/TypeOverline.vue';
import '../../assets/tailwind.css';

export default {
  title: 'Inputs/TextareaInput',
  component: TextareaInput,
  parameters: {
    componentSubtitle: 'Multi-line text input',
  },
};

const { Sizes, Variants } = Form;

export const Gallery = () => ({
  components: { TextareaInput, TypeOverline },
  data() {
    return {
      sizes: Sizes,
      variants: Variants,
    };
  },
  methods: {
    capitalize,
    handleInput: action('click'),
  },
  template: `
    <div>
      <TypeOverline tag="h1" variant="large">Textarea Inputs</TypeOverline>
      <div class="flex -mx-8">
        <div
          v-for="variant in variants"
          :key="variant"
          :class="['px-8 py-4 w-56', { 'bg-gray-0 text-white' : variant === 'ghost-inverted' }]"
        >
          <TypeOverline tag="h2" class="mb-2" variant="large">{{ variant }}</TypeOverline>
          <div v-for= "size in sizes" :key="'regular-' + variant + '-' + size" class="mb-2">
            <TypeOverline tag="h3" class="mb-2">{{ size }}, Empty</TypeOverline>
            <TextareaInput
              :variant="variant"
              :size="size"
              :placeholder="capitalize(size)"
              @input="handleInput"
            />
            <TypeOverline tag="h3" class="mb-2">{{ size }}, Filled</TypeOverline>
            <TextareaInput
              :variant="variant"
              :size="size"
              :placeholder="capitalize(size)"
              :value="capitalize(size)"
              @input="handleInput"
            />
            <TypeOverline tag="h3" class="mb-2">{{ size }}, Invalid</TypeOverline>
            <TextareaInput
              :variant="variant"
              :size="size"
              :placeholder="capitalize(size)"
              invalid
              @input="handleInput"
            />            
            <TypeOverline tag="h3" class="mb-2">{{ size }}, Disabled</TypeOverline>
            <TextareaInput
              :variant="variant"
              :size="size"
              :placeholder="capitalize(size)"
              disabled
              @input="handleInput"
            />
            <TypeOverline tag="h3" class="mb-2">{{ size }}, Disabled, Filled</TypeOverline>
            <TextareaInput
              :variant="variant"
              :size="size"
              :placeholder="capitalize(size)"
              disabled
              :value="capitalize(size)"
              @input="handleInput"
            />
          </div>
        </div>
      </div>
    </div>
  `,
});

export const WithKnobs = () => {
  return {
    components: { TextareaInput },
    props: {
      variant: {
        default: select('Variant', Object.values(Variants), Variants.STANDARD),
      },
      size: {
        default: select('Size', Object.values(Sizes), Sizes.NORMAL),
      },
      label: {
        default: text('Label', 'First name'),
      },
      placeholder: {
        default: text('Placeholder', 'Joe Smith'),
      },
      error: {
        default: text('Error', 'Something went wrong'),
      },
      invalid: {
        default: boolean('Invalid', false),
      },
    },
    methods: {
      handleInput: action('click'),
    },
    template: `
      <div class="p-8" :class="variant === 'ghost-inverted' && 'bg-gray-0'">
        <TextareaInput
          :label="label"
          :placeholder="placeholder"
          :variant="variant"
          :size="size"
          :error="error"
          :invalid="invalid"
          @input="handleInput"
        />
      </div>
    `,
  };
};

/* eslint-disable no-param-reassign */
[Gallery, WithKnobs].forEach((item) => {
  item.story = {
    parameters: {
      jest: ['TextareaInput.test.js'],
    },
  };
});
