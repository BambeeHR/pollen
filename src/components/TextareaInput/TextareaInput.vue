<template>
  <FormField
    :id="id"
    :variant="variant"
    :size="size"
    :label="label"
    :focused="hasFocus"
    :error="error"
    :invalid="invalid"
    :disabled="disabled"
    class="textarea-input"
    :class="[
      `textarea-input--${variant}`,
      `textarea-input--${size}`,
      {
        'textarea-input--invalid': invalid || error,
        'textarea-input--disabled': disabled,
      },
    ]"
  >
    <textarea
      :id="id"
      ref="input"
      :value="value"
      v-bind="$attrs"
      class="textarea-input__field"
      :aria-invalid="invalid"
      :aria-describedby="error ? `${id}-error` : false"
      :disabled="disabled"
      @input="handleInput"
      @focus="hasFocus = true"
      @blur="hasFocus = false"
    />
  </FormField>
</template>

<script>
import shortid from 'shortid';
import Form from '../../constants/Form';
import FormField from '../internal/forms/FormField.vue';

const { Sizes, Variants } = Form;

/**
 * `import { TextareInput } from '@bambee/pollen';`
 *
 * A styled `textarea` component for entering multiple lines of text. Attributes
 * not explicitly enumerated as prop (like `required`) will be transparently
 * passed through to the `textarea` element.
 */
export default {
  components: {
    FormField,
  },
  inheritAttrs: false,
  props: {
    /** If the field is disabled. */
    disabled: {
      type: Boolean,
      default: false,
    },
    /** Any messages to display as errors on the field. */
    error: {
      type: String,
      default: null,
    },
    /** A unique element ID. By default, one is randomly generated. */
    id: {
      type: String,
      default: shortid.generate,
    },
    /**
     * If true, this field will display in an error state. NOTE: a field is in
     * an error state if `invalid` is `true` and/or `error` is truthy.
     */
    invalid: {
      type: Boolean,
      default: false,
    },
    /**
     * A label to display above the field.
     */
    label: {
      type: String,
      default: null,
    },
    /**
     * One of `dense`, `regular`, and `large`.
     */
    size: {
      type: String,
      default: Sizes.NORMAL,
      validator: (value) => Object.values(Sizes).includes(value),
    },
    /** Value of the input. Compatible with `v-model`. */
    value: {
      type: String,
      default: '',
    },
    /**
     * One of `standard`, or `raised`.
     */
    variant: {
      type: String,
      default: Variants.STANDARD,
      validator: (value) => Object.values(Variants).includes(value),
    },
  },
  data() {
    return {
      hasFocus: false,
    };
  },
  methods: {
    handleInput(e) {
      this.$emit('input', e.target.value);
    },
  },
};
</script>

<style scoped>
.textarea-input__field {
  @apply appearance-none bg-transparent border-0 flex-grow font-sans;
  font-size: 1em;
}

.textarea-input__field::placeholder {
  @apply text-gray-4;
}

.textarea-input__field:focus {
  @apply outline-none;
}

/* .textarea-input--ghost-inverted .textarea-input__field {
  @apply text-white;
} */

.textarea-input--dense .textarea-input__field {
  @apply p-3;
}

.textarea-input--normal .textarea-input__field {
  @apply p-4;
}

.textarea-input--large .textarea-input__field {
  @apply p-4;
}

/* .textarea-input--ghost.textarea-input--invalid .textarea-input__field,
.textarea-input--ghost.textarea-input--invalid
  .textarea-input__field::placeholder, */
.textarea-input--raised.textarea-input--invalid .textarea-input__field,
.textarea-input--raised.textarea-input--invalid
  .textarea-input__field::placeholder {
  @apply text-error;
}

/* .textarea-input--ghost-inverted.textarea-input--invalid .textarea-input__field,
.textarea-input--ghost-inverted.textarea-input--invalid
  .textarea-input__field::placeholder {
  @apply text-error-light-ghost;
} */
</style>
