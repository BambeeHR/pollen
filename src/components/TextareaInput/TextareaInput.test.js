import { mount } from '@vue/test-utils';
import Form from '../../constants/Form';
import TextareaInput from './TextareaInput.vue';

const { Sizes, Variants } = Form;

describe('TextareaInput', () => {
  test('renders correctly', () => {
    const wrapper = mount(TextareaInput, {
      propsData: {
        variant: Variants.RAISED,
        size: Sizes.DENSE,
        error: 'There was an error',
        label: 'Test',
        id: 'test',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('transparently passes through classes and attributes', () => {
    const wrapper = mount(TextareaInput, {
      attrs: {
        id: 'test',
        disabled: true,
      },
    });
    const inputEl = wrapper.find('textarea').element;
    expect(inputEl.id).toBe('test');
    expect(inputEl.disabled).toBe(true);
  });

  test('emits an input event event', () => {
    const onInput = jest.fn();
    const wrapper = mount(TextareaInput, {
      listeners: {
        input: onInput,
      },
    });
    wrapper.find('textarea').setValue('hello world');
    expect(onInput).toHaveBeenCalledWith('hello world');
  });

  test('does not emit an input event when disabled', () => {
    const onInput = jest.fn();
    const wrapper = mount(TextareaInput, {
      listeners: {
        input: onInput,
      },
      propsData: {
        disabled: true,
      },
    });
    wrapper.find('textarea').setValue('hello world');
    expect(onInput).not.toHaveBeenCalled();
  });
});
