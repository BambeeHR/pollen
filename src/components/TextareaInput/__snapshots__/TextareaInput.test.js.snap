// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextareaInput renders correctly 1`] = `
<label
  class="form-field textarea-input textarea-input--raised textarea-input--dense textarea-input--invalid"
  for="test"
>
  <div
    class="field-label field-label--raised field-label--dense field-label--invalid"
  >
    
    Test
  
  </div>
   
  <div
    class="field-container field-container--raised field-container--dense field-container--invalid"
  >
    <textarea
      aria-describedby="test-error"
      class="textarea-input__field"
      id="test"
    />
  </div>
   
  <div
    class="field-error field-error--raised"
    id="test-error"
    role="alert"
  >
    
    There was an error
  
  </div>
</label>
`;
